lib/spack/spack/cmd/external.py

```diff
@@ -0,0 +1,194 @@
+    """Generate config according to the packages.yaml schema for a single
+    package.
+
+    This does not generate the entire packages.yaml.
+    """
+    paths_dict = syaml.syaml_dict()
+    for e in external_pkg_entries:
+        paths_dict[str(e.spec)] = e.base_dir
+    pkg_dict = syaml.syaml_dict()
+    pkg_dict['paths'] = paths_dict
+
+    pkgs_dict = syaml.syaml_dict()
+    pkgs_dict[pkg_name] = pkg_dict
+
+    return pkgs_dict
+
+
+def external_find(args):
+    _get_external_packages(TestRepo())
+
+
+def _group_by_prefix(paths):
+    groups = defaultdict(set)
+    for p in paths:
+        groups[os.path.dirname(p)].add(p)
+    return groups.items()
+
+
+def _convert_to_iterable(single_val_or_multiple):
+    x = single_val_or_multiple
+    if x is None:
+        return []
+    elif isinstance(x, six.string_types):
+        return [x]
+    elif isinstance(x, spack.spec.Spec):
+        # Specs are iterable, but a single spec should be converted to a list
+        return [x]
+
+    try:
+        iter(x)
+        return x
+    except TypeError:
+        return [x]
+
+
+def _determine_base_dir(prefix):
+    assert os.path.isdir(prefix)
+    if os.path.basename(prefix) == 'bin':
+        return os.path.dirname(prefix)
+
+
+def _get_external_packages(repo, system_path_to_exe=None):
+    if not system_path_to_exe:
+        system_path_to_exe = _get_system_executables()
+
+    exe_pattern_to_pkgs = defaultdict(list)
+    for pkg in repo.all_packages():
+        if hasattr(pkg, 'executables'):
+            for exe in pkg.executables:
+                exe_pattern_to_pkgs[exe].append(pkg)
+
+    pkg_to_found_exes = defaultdict(set)
+    for path, exe in system_path_to_exe.items():
+        for exe_pattern, pkgs in exe_pattern_to_pkgs.items():
+            if re.search(exe_pattern, exe):
+                for pkg in pkgs:
+                    pkg_to_found_exes[pkg].add(path)
+
+    pkg_to_entries = defaultdict(list)
+    resolved_specs = {}  # spec -> exe found for the spec
+
+    # TODO: iterate through this in a predetermined order (e.g. by package
+    # name) to get repeatable results when there are conflicts.
+    for pkg, exes in pkg_to_found_exes.items():
+        for prefix, exes_in_prefix in _group_by_prefix(exes):
+            # TODO: multiple instances of a package can live in the same
+            # prefix, and a package implementation can return multiple specs
+            # for one prefix, but without additional details (e.g. about the
+            # naming scheme which differentiates them), the spec won't be
+            # usable.
+            if hasattr(pkg, 'determine_spec_details'):
+                specs = _convert_to_iterable(
+                    pkg.determine_spec_details(prefix, exes_in_prefix))
+            else:
+                # Note: packages which do not implement
+                # 'determine_spec_details' should use exact-match regular
+                # expressions
+                specs = [spack.spec.Spec(pkg.name)]
```

Comment:
Just thinking loud about this, but I wonder if we should mandate for packages that implement this protocol something like `determine_version` and then have `determine_spec_details` act on executables that are grouped by prefix AND version.

The idea is that the regex pattern will be used to find matching executables, then:
1. We group by prefix
2. We group each prefix by version
3. We call `determine_spec_details(prefix, versions, exes)`

Reply:
To make sure I understand: is this to handle cases where multiple versions can be found in the same prefix?

I was originally thinking we could provide a utility function to start with an identifier string (like "gcc") and then return all the prefix/suffix combinations (and also possibly filter them based on whether they look like versions i.e. to avoid pulling in a file like `gccmakedep` and assuming it has a version `makedep`). The user could call this inside of their `determine_spec_details` implementation

I think it may be advantageous to allow the user to inspect candidate files more closely. It does create some extra work by requiring them to call the function. I think your suggestion may be better (since they can test the suggested exe after the fact).

My only thought is that it would be good to allow the user to disable this for some packages, possibly in the config and/or possibly at the package level.

Reply:
> To make sure I understand: is this to handle cases where multiple versions can be found in the same prefix?

Yes. More in general I wonder if we should always require that a version is detected. 

> My only thought is that it would be good to allow the user to disable this for some packages, possibly in the config and/or possibly at the package level

Do you think it is possible to come up with a protocol similar to that of build-systems? There each build system has "phases" which usually have a default implementation that permit users to implement some callback to influence them. Taking as an example `Autotools` what most packages need to do is to implement `configure_args` and rely on the defaults of the base class for phases. If need be though, packages can completely override one or more phases and achieve more generality.

Reply:
> I think your suggestion may be better (since they can test the suggested exe after the fact).

I was thinking about your suggestion more and I'm actually not sure if it is any easier for the user to do things this way. Right now, the compiler logic groups executables by suffixes/prefixes that are extracted from the name (in make_compiler_list), but I’m wondering if this is necessary (for compilers or for detecting external packages):

If there are many instances of an installed package in a prefix, we will have to invoke each of the executables to determine what spec they are associated with, regardless of the naming scheme. Therefore, we can group executables by spec with or without being able to extract prefixes/suffixes from the naming scheme.

At least, this works well enough for versions, a naming scheme may provide more details than just the version, and it may be more difficult to learn these details by invoking an executable. I think a utility function could be provided so that a user could group all the executables by prefix/suffix with respect to an anchor string though in this case.

> I wonder if we should always require that a version is detected.

When examining packages with executables, the user should be able to verify that the output matches what they expect, so generally I agree that it should be possible to detect a version. I think that means that any package which defines `executables` must also define `determine_spec_details` (right now that’s optional) - do you agree?

> Do you think it is possible to come up with a protocol similar to that of build-systems? There each build system has "phases" which usually have a default implementation that permit users to implement some callback to influence them. 

I think that works well for build systems because they are established. My inclination is to provide utility functions to avoid constraining the user in case their package doesn't conform to our expectations

Reply:
>  I think that means that any package which defines executables must also define determine_spec_details (right now that’s optional) - do you agree?

I agree with this.  I think we should encourage packages that try to do this to provide more information.  We *could* define a default `determine_spec_details` that invokes with `--version` or something, but I'm hesitant to do that as there is such a wide range of behaviors that executables can have, and even if `--version` works, every command has a different syntax.  So yes, require it.

On specs vs. versions, etc.: I like the idea of detecting specs, then grouping similar specs, because it's more general than what the compiler logic currently does.  The compilers currently group by prefix, suffix, and version, but that is very specialized. The prefixes and suffixes are really for toolchains or variants -- e.g., what arch is the compiler for, or what was it built by (e.g., `-mp` for `macports`).  I think that stuff will generally come out as a version suffix or as a variant on the spec with this new logic.  It also simplifies the code -- the compiler code passes around a bunch of tuples in weird ways to various threaded routines.  This could generate specs first and consolidate the grouping logic in one place.

Only thing is we do not currently have a "general" way to group specs with similar attributes.  I guess there is `llnl.util.lang.index_by` -- it's used in some of the compiler commands. But then again, compilers only group things to try to match corresponding cc/cxx/fortran executables together.  We wouldn't need that yet.  Here, we are just saying "this prefix contains these specs", so if you have the specs do you need much else?

Reply:
> so if you have the specs do you need much else?

No, but my thought was there may be occasions where the `package.py`'s `determine_spec_details` implementation would have trouble determining the full spec, and may have to depend on the name. In that case, I think being able to parse out prefixes/suffixes would be useful, but not necessary for an initial cut.

Reply:
> If there are many instances of an installed package in a prefix, we will have to invoke each of the executables to determine what spec they are associated with, regardless of the naming scheme. Therefore, we can group executables by spec with or without being able to extract prefixes/suffixes from the naming scheme.

I wonder if there are niche cases where this is not true. For compilers a common prefix/suffix usually indicates that executables belong to the same spec. In any case thinking about how this PR could evolve later I don't see us cornered in any way. The API in `determine_spec_details` is general enough to permit whatever refinement in its implementation.

As a point for later discussion, we might think if it makes sense to design different mixins depending on the type of package. Each mixin might provide a default implementation for `determine_spec_details` and require to implement more specific methods with a narrower scope e.g. `determine_version` or whatever makes sense for that package type.

> I think that means that any package which defines executables must also define determine_spec_details (right now that’s optional) - do you agree?

+1 on this

> This could generate specs first and consolidate the grouping logic in one place.

+1 also to this